2-Draft-Accepted/RFCxxxx-PowerShellGet-3.0.md

```diff
@@ -0,0 +1,242 @@
+
+### Automatic updating of the cache
+
+On any operation that reaches out to a repository, a REST API will be called to
+see if the hash of the cache matches the current cache and if not, a new one
+is downloaded.
+If the repository doesn't support this new API, it falls back to current behavior
+in PSGet v2.
+
+### Repository management
+
+`Register-PSResourceRepository` will allow for registering additional repositories.
+A `-Default` switch enables registering PSGallery should it be accidentally removed.
+The `-URL` will accept the HTTP address without the need to specify `/api/v2` as
+that will be assumed and discovered at runtime.
+Support for local filesystem repositories must be maintained.
+A `-InstallationPolicy` switch accepts `Trusted` or `Untrusted` (default) indicating
+whether to prompt the user when installing resources from that repository.
+
+`Get-PSResourceRepository` will list out the registered repositories.
+
+`Set-PSResourceRepository` can be used to update a repository URL or installation policy.
+
+`Unregister-PSResourceRepository` can be used to un-register a repository.
+
+### Finding resources
+
+Current design of PowerShellGet is to have different cmdlets for each of the
+different resources types that it supports:
+
+`Find-Command`
+`Find-DscResource`
+`Find-Module`
+`Find-RoleCapability`
+`Find-Script`
+
+Instead, these will be abstracted by a `Find-PSResource` cmdlet.
+A `-Type` parameter will accept an array of types to allow filtering.
+
+With support of the generic `PSResource`, this means we can also find and
+install arbitrary nupkgs that may only contain an assembly the user wants to
+use in their script.
+
+If the local cache does not exist, the cmdlet will call `Update-PSResourceCache`
+first and then attempt a search.
+However, this cmdlet does not update the cache if one exists.
+
+### Installing resources
+
+`Install-PSResource` will only work for modules.
+A `-AllowPrerelease` switch allows installing prerelease versions.
+Other types will use `Save-PSResource` (see below).
+A `-Repository` parameter accepts a specific repository name.
+If `-Repository` is not specified and the same resource is found in multiple
+repositories, then the resource is automatically installed quietly from the
+trusted repository with the highest version matching the `-Version` parameter
+(if specified, otherwise newest non-prerelease version unless `-AllowPrerelease`
+is used).
+If there are no trusted repositories matching the query, then the newest version
+fulfilling the query will be prompted to be installed.
+If there are multiple repositories with the same trust level containing the same
+version, the first one is used.
+
+`-AllowUntrusted` can be used to suppress being prompted for untrusted sources.
+`-AllowDifferentPublisher` can be used to suppress being prompted if the publisher
+of the module is different from the currently installed version.
+`-AllowReinstall` can be used to re-install a module.
```

Comment:
Can we try to find parameter names that avoid having a bunch that start with the same 5-character prefix?

Suggestions:
- `-IgnorePublisher`
- `-TrustRepository`
- `-Reinstall`

Reply:
I would also add `-Prerelease` to that list.

Reply:
Each `-Allow*` would have an alias `-au`, `-ad`, `-ar`.  Having the same prefix groups them together making it easier for newer users to understand (I think).  Will discuss with @PowerShell/powershell-committee as I believe current convention is `-Allow*`

Reply:
Aliases do not help as much with infrequent commands (tab-completion doesn't expand them, so if I'm not 100% sure I remember correctly, I can't use them), I end up {Tab}{Tab}{Tab}{Tab}{Tab} to cycle through them all and then back around to the one I wanted.

Also, let's not forget that the multiple `-Allow` parameters are a **painful** convention that users have been complaining about since the first release of PowerShellGet -- let's not forget the #1 upvoted request on PowerShellGet is about _exactly_ the proliferation of these "yes, really, just do it" parameters ...

Reply:
PS 6.2.1 has 10 cmdlets with parameters using the prefix `-Allow` and 3 unique parameters `-AllowRedirection`, `-AllowClobber`, `AllowUnencryptedAuthenication`. It's not a widely used prefix and I would say it actually hurts the user experience when there are multiple parameters on the same cmdlet using the same prefix. Users that prefer tabbing rather than aliases requires more key presses to get to the desired parameter. Inexperienced users are much more likely to tab through parameters than knowing an `-au` alias exists for the parameter.

Semantically `-AllowReinstall` doesn't make sense. The user is telling the cmdlet to perform a reinstall so a `-Reinstall` parameter is easier for users rather than thinking, "Why am I granting the cmdlet permission to do a reinstall and in what cases does it perform a reinstall?" with `-AllowReinstall`.

Reply:
Got the feedback, will go with unique ones.